sqs_workers/memory_sqs.py

```diff
@@ -158,19 +160,40 @@ def delete_messages(self, Entries):
         message_ids = {entry["Id"] for entry in Entries}
 
         successfully_deleted = set()
-        push_back_messages = []
 
-        for message in self.messages:
+        for i, message in enumerate(self.in_flight):
             if message.message_id in message_ids:
                 successfully_deleted.add(message.message_id)
-            else:
-                push_back_messages.append(message)
-        self.messages[:] = push_back_messages
+                del self.in_flight[i]
 
-        didnt_deleted = message_ids.difference(successfully_deleted)
         return {
             "Successful": [{"Id": _id} for _id in successfully_deleted],
-            "Failed": [{"Id": _id} for _id in didnt_deleted],
+        }
+
+    def change_message_visibility_batch(self, Entries):
+        """
+        Changes message visibility by looking at in-flight messages, setting
+        a new execute_at, and returning it to the pool of processable messages
+        """
+        edited = []
+        return_to_pool = []
+        entries_by_id = {e["Id"]: e for e in Entries}
+
+        for i, m in enumerate(self.in_flight):
+            if m.message_id in entries_by_id.keys():
+                sec = int(entries_by_id[m.message_id]["VisibilityTimeout"])
+                now = datetime.datetime.utcnow()
+                execute_at = now + datetime.timedelta(seconds=sec)
+                changed = attr.evolve(m, execute_at=execute_at)
+                changed.attributes["ApproximateReceiveCount"] += 1
+                edited.append(changed)
+                return_to_pool.append(changed)
+                del self.in_flight[i]
+
+        self.messages.extend(return_to_pool)
+
+        return {
+            "Successful": [{"Id": _id} for _id in edited],
```

Comment:
Here too, I might be missing something.

Here for `change_message_visibility_batch` and near [L170](https://github.com/Doist/sqs-workers/pull/98/files#diff-deb4c408dd0cf806b3af78ce4651624a3986aa2b66ccf9eafeb092747fac9c02R170) for `deleted_messages` we stopped returning which messages failed, yet in lines [L153](https://github.com/Doist/sqs-workers/pull/98/files#diff-340590a9761a20e6d96d567e9dabee10bdd0a4de28a3d210e05a221efff0b1fbR153), [L163](https://github.com/Doist/sqs-workers/pull/98/files#diff-340590a9761a20e6d96d567e9dabee10bdd0a4de28a3d210e05a221efff0b1fbR163) and [L232](https://github.com/Doist/sqs-workers/pull/98/files#diff-340590a9761a20e6d96d567e9dabee10bdd0a4de28a3d210e05a221efff0b1fbR233) we check for them. Won't this always succeed, since the `Failed` key is now absent?

Reply:
Ah this is a good thing to ask!

So, at least for `delete_messages`, the current impl is totally broken (see my other [comment](https://github.com/Doist/sqs-workers/pull/98#discussion_r1697846864)), and it doesn't seem to matter, so I figured I'd just simplify as I fixed it, right.

But, looking now, I think the `Failed` key for the response of `delete_messages` can still be meaningful (it can return the message IDs which weren't found in `in_flight`, which makes enough sense), so I'll add that

As for `change_message_visibility_batch` -- I'm not sure what a "failed" message would look like! Maybe similar to the above, where it can represent the set of message IDs which weren't found in `in_flight`?

> Won't this always succeed, since the Failed key is now absent?

As for this -- yes! But my understanding was that it didn't really matter as long as it works for tests, since that's all that `MemoryQueue` is used for (at least, I dearly hope, lol, since it's so broken). Still, no reason to not improve it like we're talking about here, so I'll make some updates 🙂

Reply:
This is done!

Reply:
> As for change_message_visibility_batch -- I'm not sure what a "failed" message would look like! Maybe similar to the above, where it can represent the set of message IDs which weren't found in in_flight?

Yeah, this makes sense to me.

> But my understanding was that it didn't really matter as long as it works for tests, since that's all that MemoryQueue is used for (at least, I dearly hope, lol, since it's so broken). Still, no reason to not improve it like we're talking about here, so I'll make some updates 🙂

Right! As said below, I didn't realize this was only used in tests, so my commentary didn't account for that. I agree though that these are nice improvements. Thanks for making them!